File: lib/backend/src/utils/paths.ts
import path from 'path';
import os from 'os';
import fs from 'fs';

function getAppDir(): string {
  const home = os.homedir();
  if (process.platform === 'win32') {
    return path.join(process.env.APPDATA || path.join(home, 'AppData', 'Roaming'), 'LuyumiLauncher');
  } else if (process.platform === 'darwin') {
    return path.join(home, 'Library', 'Application Support', 'LuyumiLauncher');
  } else {
    return path.join(home, '.luyumilauncher');
  }
}

// Initialize with IIFE to ensure synchronous execution
export const APP_DIR = (() => getAppDir())();
export const CACHE_DIR = (() => path.join(APP_DIR, 'cache'))();
export const TOOLS_DIR = (() => path.join(APP_DIR, 'butler'))();
export const GAME_DIR = (() => path.join(APP_DIR, 'install', 'release', 'package', 'game', 'latest'))();
export const JRE_DIR = (() => path.join(APP_DIR, 'install', 'release', 'package', 'jre', 'latest'))();
export const PROFILES_DIR = (() => path.join(APP_DIR, 'profiles'))();

export function getResolvedAppDir(customPath?: string): string {
  if (customPath && customPath.trim()) {
    return path.join(customPath.trim(), 'LuyumiLauncher');
  }
  try {
    const configFile = path.join(APP_DIR, 'config.json');
    if (fs.existsSync(configFile)) {
      const config = JSON.parse(fs.readFileSync(configFile, 'utf8'));
      if (config.installPath && config.installPath.trim()) {
        return path.join(config.installPath.trim(), 'LuyumiLauncher');
      }
    }
  } catch (err) {
    // Ignore config errors, use default
  }
  return APP_DIR;
}

export function expandHome(inputPath: string | null | undefined): string | null {
  if (!inputPath) {
    return null;
  }
  if (inputPath === '~') {
    return os.homedir();
  }
  if (inputPath.startsWith('~/') || inputPath.startsWith('~\\')) {
    return path.join(os.homedir(), inputPath.slice(2));
  }
  return inputPath;
}

export function getClientCandidates(gameLatest: string) {
  const candidates: string[] = [];
  if (process.platform === 'win32') {
    // Windows - Only HytaleClient.exe
    candidates.push(path.join(gameLatest, 'Client', 'HytaleClient.exe'));
  } else if (process.platform === 'darwin') {
    // macOS
    candidates.push(path.join(gameLatest, 'Client', 'Hytale.app', 'Contents', 'MacOS', 'HytaleClient'));
    candidates.push(path.join(gameLatest, 'Client', 'HytaleClient'));
    candidates.push(path.join(gameLatest, 'Hytale.app', 'Contents', 'MacOS', 'HytaleClient'));
  } else {
    // Linux
    candidates.push(path.join(gameLatest, 'Client', 'HytaleClient'));
    candidates.push(path.join(gameLatest, 'HytaleClient'));
  }
  return candidates;
}

export function findUserDataPath(gameLatest: string): string | null {
  function searchDirectory(dir: string): string | null {
    try {
      const items = fs.readdirSync(dir, { withFileTypes: true });
      for (const item of items) {
        if (item.isDirectory()) {
          const fullPath = path.join(dir, item.name);
          if (item.name === 'UserData') {
            return fullPath;
          }
          // Limit recursion depth if needed, or rely on game structure being shallow enough
          const found = searchDirectory(fullPath);
          if (found) return found;
        }
      }
    } catch (error) {
      // Ignore access errors
    }
    return null;
  }

  if (!fs.existsSync(gameLatest)) {
    return null;
  }

  // First check common locations to avoid deep scan
  const candidates = [
    path.join(gameLatest, 'UserData'),
    path.join(gameLatest, 'Client', 'UserData')
  ];

  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) return candidate;
  }

  return searchDirectory(gameLatest);
}

export function getGameModsPath(): string {
  // Try to find UserData
  const userData = findUserDataPath(GAME_DIR);
  if (userData) {
    return path.join(userData, 'Mods');
  }
  // Fallback if UserData not found (e.g. game not installed yet)
  return path.join(GAME_DIR, 'Client', 'UserData', 'Mods');
}

export function findClientPath(gameLatest: string) {
  const candidates = getClientCandidates(gameLatest);
  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) {
      return candidate;
    }
  }
  return null;
}

File: lib/backend/src/utils/paths.py
import os
import sys
import json
from .platform import is_windows, is_mac, is_linux

def get_app_dir():
    app_name = "LuyumiLauncher"
    if is_windows():
        base = os.environ.get('APPDATA')
        if not base:
            base = os.path.join(os.path.expanduser('~'), 'AppData', 'Roaming')
        return os.path.join(base, app_name)
    elif is_mac():
        return os.path.join(os.path.expanduser('~'), 'Library', 'Application Support', app_name)
    elif is_linux():
        return os.path.join(os.path.expanduser('~'), '.local', 'share', app_name)
    else:
        return os.path.join(os.path.expanduser('~'), '.config', app_name)

def get_resolved_app_dir():
    app_name = "LuyumiLauncher"
    app_dir = get_app_dir()
    config_paths = [os.path.join(app_dir, 'config.json')]
    
    if is_windows():
        roaming = os.environ.get('APPDATA')
        if roaming:
            roaming_dir = os.path.join(roaming, app_name)
            if roaming_dir != app_dir:
                config_paths.append(os.path.join(roaming_dir, 'config.json'))
    
    for config_path in config_paths:
        try:
            if os.path.exists(config_path):
                with open(config_path, 'r', encoding='utf-8') as f:
                    config = json.load(f)
                install_path = config.get('installPath')
                if install_path and str(install_path).strip():
                    return os.path.join(str(install_path).strip(), app_name)
        except Exception:
            pass
    
    return app_dir

def expand_home(path_str: str):
    if not path_str:
        return path_str
    return os.path.expanduser(path_str)

def get_profiles_dir():
    return os.path.join(get_resolved_app_dir(), 'profiles')

def find_client_path(game_dir: str):
    candidates = [
        os.path.join(game_dir, "Hytale.exe"),
        os.path.join(game_dir, "Client", "Hytale.exe"),
        os.path.join(game_dir, "HytaleClient.exe"),
        os.path.join(game_dir, "Client", "HytaleClient.exe"),
        # Linux/Mac candidates
        os.path.join(game_dir, "HytaleClient"),
        os.path.join(game_dir, "Client", "HytaleClient"),
    ]
    for c in candidates:
        if os.path.exists(c):
            return c
    return None

def get_game_dir():
    return os.path.join(get_resolved_app_dir(), 'Hytale')

def get_game_mods_path():
    return os.path.join(get_game_dir(), 'mods')

def get_user_data_dir():
    r"""
    Get the UserData directory for the launcher.
    Always returns the launcher's base directory UserData, not affected by custom install paths.
    
    Windows: C:\Users\{user}\AppData\Roaming\LuyumiLauncher\userData
    Linux: ~/.local/share/LuyumiLauncher/userData
    macOS: ~/Library/Application Support/LuyumiLauncher/userData
    """
    return os.path.join(get_app_dir(), 'userData')
